### ISSUE TITLE
`get_objects_for_user()` executes insane `__in` queries

### ISSUE THREAD

**Issue by jnm** [kobotoolbox/kpi#2671]
OCHA database server being hammered at peak times by queries like:
```sql
SELECT COUNT (*)
FROM
  (SELECT DISTINCT "kpi_asset"."id" AS Col1,
                   "kpi_asset"."name" AS Col2,
                   "kpi_asset"."date_created" AS Col3,
                   "kpi_asset"."date_modified" AS Col4,
                   "kpi_asset"."summary" AS Col5,
                   "kpi_asset"."report_custom" AS Col6,
                   "kpi_asset"."map_styles" AS Col7,
                   "kpi_asset"."map_custom" AS Col8,
                   "kpi_asset"."asset_type" AS Col9,
                   "kpi_asset"."parent_id" AS Col10,
                   "kpi_asset"."owner_id" AS Col11,
                   "kpi_asset"."editors_can_change_permissions" AS Col12,
                   "kpi_asset"."uid" AS Col13,
                   "kpi_asset"."settings" AS Col14,
                   "kpi_asset"."_deployment_data" AS Col15
   FROM "kpi_asset"
   LEFT OUTER JOIN "kpi_collection" ON (
      "kpi_asset"."parent_id" = "kpi_collection"."id")
   LEFT OUTER JOIN "kpi_usercollectionsubscription" ON (
      "kpi_collection"."id" = "kpi_usercollectionsubscription"."collection_id")
   WHERE ("kpi_asset"."id" IN (1886499, 1886520, 1886671, 1931865,
                                 1941583, 1960253, 1960254, 2008205,
                                 2029000, 2029026, 2029027, 2065716,
                                 2103754, 2104743, 2104914, 2104974,
                                 2113470, 2146030, 2215903, 2217720,
                                 2351689, 2369900, 2406365, 2406520,
                                 2406557, 2411373, 2425261, 2429947,
                                 2464191, 2478116)
          OR ("kpi_usercollectionsubscription"."user_id" = 63836
              AND "kpi_asset"."id" IN (XXXXXXX))) ) subquery
```
…where `XXXXXXX` is _over twenty-two thousand_ IDs, or basically the ID of every publicly-accessible asset.

These two calls execute the same enormous `__in` query:
```python
get_objects_for_user(
    # Asset.objects.all().count() is over 600K
    get_anonymous_user(), 'kpi.view_asset', Asset.objects.all()
).count()
```
```python
get_objects_for_user(
    get_anonymous_user(), 'kpi.view_asset',
    Asset.objects.filter(pk__lt=100)
).count()
```

Once #2332 is merged into `master`, then all this imitation-django-guardian, generic-relation stuff can be put out to pasture for good—but right now we need immediate relief.

**Comment by jnm** [comment]
Fix for old releases running on OCHA and HHI: https://github.com/kobotoolbox/kpi/releases/tag/two-databases_2.020.12a
- [x] Spin up old environment to run unit tests
- [x] Deploy to HHI
- [ ] Deploy to OCHA
- [ ] Make PR to `master` instead of archaic stuff

**Comment by jnm** [comment]
Drat, the "fix" doesn't work with https://github.com/kobotoolbox/kpi/blob/9c70d5062f0e7bc2c7650ea93d1d43d72ff19b84/kobo/apps/reports/views.py#L27-L34

uWSGI `HARAKIRI`s abound…

**Comment by jnm** [comment]
Kludge for the autoreport: https://github.com/kobotoolbox/kpi/releases/tag/two-databases_2.020.12b / https://github.com/kobotoolbox/kpi/commit/f9ae3aeabc164eaf268b5b129c58ee91c58ce568

**Comment by jnm** [comment]
Before things went sour, the good news was (before and after):
![image](https://user-images.githubusercontent.com/2085013/81901636-d3d2f780-958c-11ea-82a4-649755f7167d.png)

**Comment by jnm** [comment]
Have to consider these places where `get_objects_for_user()` is called:
https://github.com/kobotoolbox/kpi/blob/1acf9a54ae80816d5b2c5f087fdde45d4e3118d1/kpi/views/v1/tag.py#L42-L49
https://github.com/kobotoolbox/kpi/blob/1acf9a54ae80816d5b2c5f087fdde45d4e3118d1/kpi/views/v2/asset.py#L377-L379
https://github.com/kobotoolbox/kpi/blob/1acf9a54ae80816d5b2c5f087fdde45d4e3118d1/kpi/serializers/v1/user_collection_subscription.py#L25-L29
https://github.com/kobotoolbox/kpi/blob/1acf9a54ae80816d5b2c5f087fdde45d4e3118d1/kpi/filters.py#L53-L62
https://github.com/kobotoolbox/kpi/blob/1acf9a54ae80816d5b2c5f087fdde45d4e3118d1/kobo/apps/reports/views.py#L31-L37
